code/game/machinery/vending/cm_vending.dm

```diff
@@ -902,50 +905,133 @@ IN_USE						used for vending/denying
 	. = ..()
 	populate_product_list(1.2)
 
+	for (var/i in 1 to length(listed_products))
+		// initial item count setup
+		var/item_name = listed_products[i][1]
+		var/initial_count = listed_products[i][2]
+		initial_product_count[item_name] = initial_count
+
+		// icon setup
+		var/typepath = listed_products[i][3]
+		var/obj/item/I = typepath
+		var/icon_ref = initial(I.icon)
+		var/icon_state = initial(I.icon_state)
+		var/map_decor = initial(I.map_specific_decoration)
+		if (map_decor)
+			icon_ref = "icons/obj/items/weapons/guns/guns_by_map/classic/guns_obj.dmi"
+		var/icon/r = icon(icon_ref, icon_state, SOUTH, 1)
+		var/result = icon2html(r, world, icon_state, sourceonly=TRUE)
+
+		product_icon_list[item_name] = list(
+			"href"=result
+		)
+
 //this proc, well, populates product list based on roundstart amount of players
 /obj/structure/machinery/cm_vending/sorted/proc/populate_product_list(var/scale)
 	return
 
+/obj/structure/machinery/cm_vending/sorted/ui_state(mob/user)
+	return GLOB.not_incapacitated_and_adjacent_state
+
+/obj/structure/machinery/cm_vending/sorted/ui_status(mob/user, datum/ui_state/state)
```

Comment:
I think some of these vendors also have access requirements? therefore you should also check allowed() here

Reply:
I've refactored the allowed / hacked etc access checks and call it in here - should resolve this properly